src/Windows/R/Components/Impl/InteractiveWorkflow/Implementation/InteractiveWindowWriter.cs

```diff
@@ -0,0 +1,198 @@
+
+            Debug.Assert(_currentLineStart >= 0);
+            if (_currentLineStart >= 0) {
+                var snapshot = _interactiveWindow.OutputBuffer.CurrentSnapshot;
+
+                var span = Span.FromBounds(_currentLineStart, snapshot.Length);
+                var currentText = snapshot.GetText(span);
+
+                var index = GetDifference(currentText, _newText, out string difference);
+                if (index >= 0) {
+                    span = Span.FromBounds(_currentLineStart + index, snapshot.Length);
+                    _interactiveWindow.OutputBuffer.Replace(span, difference);
+                } else {
+                    _interactiveWindow.OutputBuffer.Replace(span, _newText);
+                }
+            }
+        }
+
+        private static int GetDifference(string oldText, string newText, out string difference) {
+            difference = string.Empty;
+            if (newText.Length < oldText.Length) {
+                return -1;
+            }
+
+            var index = -1;
+
+            for (var i = 0; i < oldText.Length; i++) {
+                if (oldText[i] != newText[i]) {
+                    index = i;
+                    break;
+                }
+            }
+
+            difference = newText.Substring(index);
+            return index;
+        }
+
+        internal class Message {
+            public string Text;
+            public bool IsError;
+
+            public bool MovesCaretBack() => Text.Length > 1 && Text[0] == '\r' && Text[1] != '\n';
+            public bool ContainsLineFeed() => Text.Any(ch => ch == '\n');
+            public bool IsPlainText() => !MovesCaretBack() && !ContainsLineFeed();
+        }
+
+        internal class MessageQueue : IDisposable {
+            private readonly List<Message> _messages = new List<Message>();
+            private readonly ManualResetEventSlim _messagesAvailable = new ManualResetEventSlim(false);
+            private readonly object _lock = new object();
+            private int _lastCRMessageIndex = -1;
+
+            public void Enqueue(string text, bool isError) {
+                lock (_lock) {
+                    var m = new Message { Text = text, IsError = isError };
+                    // Try merging text with the last CR message, if any
+                    if (_messages.Count > 0) {
+                        var lastMessage = _messages[_messages.Count - 1];
+                        if (lastMessage.MovesCaretBack() && m.IsPlainText() && lastMessage.IsError == m.IsError) {
+                            lastMessage.Text += text;
+                            return;
+                        }
+                    }
+
+                    if (m.MovesCaretBack()) {
+                        // CR moves caret back effectively erasing previous message at the line
+                        if (_lastCRMessageIndex >= 0) {
+                            _messages.RemoveRange(_lastCRMessageIndex, _messages.Count - _lastCRMessageIndex);
```

Comment:
Also, shouldn't we remove only part of message if it is longer than one character?